File: adnetworks/Afio/MovieReward6010.m
//
//
//  MovieReward6010.m(Afio)
//
//  Copyright (c) A .D F. U. L. L. Y Co., Ltd. All rights reserved.
//
//

#import "MovieReward6010.h"

@interface MovieReward6010()
@property (nonatomic, strong) NSString *sid;
@property (nonatomic, strong) NSString *tag;
@property (nonatomic) BOOL didLoad;
@end

@implementation MovieReward6010
/**
 *  データの設定
 */
-(void)setData:(NSDictionary *)data {
    [super setData:data];
    
    self.sid = [data objectForKey:@"sid"];
    self.tag = [data objectForKey:@"tag"];
    if (!self.tag || [self.tag isEqual: [NSNull null]]){
        self.tag = @"";
    }
    
    self.didLoad = NO;
}

-(void)initAdnetworkIfNeeded {
    if (!self.amoadInterstitialVideo) {
        self.amoadInterstitialVideo = [AMoAdInterstitialVideo sharedInstanceWithSid:self.sid tag:self.tag];
        [self.amoadInterstitialVideo setDelegate:self];
        [self setCancellable];
    }
}

/**
 *  広告の読み込みを開始する
 */
-(void)startAd {
    // 動画の読み込みを開始します。
    if (self.amoadInterstitialVideo && !self.amoadInterstitialVideo.isLoaded) {
        [self.amoadInterstitialVideo load];
        self.didLoad = YES;
    }
}

-(BOOL)isPrepared {
    if (self.didLoad && self.delegate && self.amoadInterstitialVideo && self.amoadInterstitialVideo.isLoaded) {
        return YES;
    }
    return NO;
}

/**
 *  広告の表示を行う
 */
-(void)showAd {
    [super showAd];
    
    if ([self isPrepared]) {
        [self.amoadInterstitialVideo show];
        self.didLoad = NO;
    } else {
        [self setCallbackStatus:MovieRewardCallbackPlayFail];
    }
}
-(void)showAdWithPresentingViewController:(UIViewController *)viewController {
    [self showAd];
}

/**
 * 対象のクラスがあるかどうか？
 */
-(BOOL)isClassReference {
    Class clazz = NSClassFromString(@"AMoAdInterstitialVideo");
    if (clazz) {
    } else {
        NSLog(@"Not found Class: AMoAdInterstitialVideo");
        return NO;
    }
    return YES;
}

-(void)setCancellable {
    if (self.amoadInterstitialVideo) {
        [self.amoadInterstitialVideo setCancellable:NO];
    }
}

#pragma mark - AMoAdInterstitialVideoDelegate
- (void)amoadInterstitialVideo:(AMoAdInterstitialVideo *)amoadInterstitialVideo didLoadAd:(AMoAdResult)result {
    // 広告のロードが完了した
    NSLog(@"%s", __FUNCTION__);
    if (result == AMoAdResultSuccess) {
        NSLog(@"%s %@", __FUNCTION__, @"AMoAdResultSuccess");
        [self setCallbackStatus:MovieRewardCallbackFetchComplete];
    } else {
        [self setCallbackStatus:MovieRewardCallbackFetchFail];
    }
    
}

- (void)amoadInterstitialVideoDidStart:(AMoAdInterstitialVideo *)amoadInterstitialVideo {
    // 動画の再生を開始した
    NSLog(@"%s", __FUNCTION__);
    [self setCallbackStatus:MovieRewardCallbackPlayStart];
}

- (void)amoadInterstitialVideoDidComplete:(AMoAdInterstitialVideo *)amoadInterstitialVideo {
    // 動画を最後まで再生完了した
    NSLog(@"%s", __FUNCTION__);
    [self setCallbackStatus:MovieRewardCallbackPlayComplete];
}

- (void)amoadInterstitialVideoDidFailToPlay:(AMoAdInterstitialVideo *)amoadInterstitialVideo {
    // 動画の再生に失敗した
    NSLog(@"%s", __FUNCTION__);
    [self setCallbackStatus:MovieRewardCallbackPlayFail];
}

- (void)amoadInterstitialVideoDidShow:(AMoAdInterstitialVideo *)amoadInterstitialVideo {
    // 広告を表示した
    NSLog(@"%s", __FUNCTION__);
}

- (void)amoadInterstitialVideoWillDismiss:(AMoAdInterstitialVideo *)amoadInterstitialVideo {
    // 広告を閉じた
    NSLog(@"%s", __FUNCTION__);
    [self setCallbackStatus:MovieRewardCallbackClose];
}

- (void)amoadInterstitialVideoDidClickAd:(AMoAdInterstitialVideo *)amoadInterstitialVideo {
    // 広告がクリックされた
    NSLog(@"%s", __FUNCTION__);
}
@end

File: adnetworks/Afio/MovieReward6010.swift
//
//  MovieReward6010.swift
//  MovieRewardTestApp
//
//  Created by Sungil Kim on 2020/05/20.
//  Copyright © 2020 Glossom, Inc. All rights reserved.
//

import Foundation
import AMoAd
import ADFMovieReward

@objc(MovieReward6010)

class MovieReward6010: ADFmyMovieRewardInterface {
    var amoadInterstitialVideo: AMoAdInterstitialVideo?
    private var didLoad = false
    
    override init() {
        super.init()
        configure = AdnetworkConfigure6010.sharedInstance()
    }

    override class func getAdapterRevisionVersion() -> String {
        return "8"
    }
    
    override class func adnetworkClassName() -> String {
        return "AMoAd.AMoAdInterstitialVideo"
    }
    
    override class func adnetworkName() -> String {
        return AdnetworkConfigure6010.adnetworkName()
    }
    
    override func setData(_ data: [AnyHashable : Any]) {
        print("MovieReward6010: setData")
        super.setData(data)
        
        adParam = AdnetworkParam6010.init(param: data)
        configure.param = adParam
    }

    override func initAdnetworkIfNeeded() -> Bool {
        guard super.initAdnetworkIfNeeded() == true else {
            return false
        }
        
        weak var weakSelf = self
        configure.initAdnetworkSDK { result in
            guard let strongSelf = weakSelf else {
                return
            }
            guard strongSelf.amoadInterstitialVideo == nil, let param = strongSelf.adParam as? AdnetworkParam6010, let sid = param.sid else {
                print("MovieReward6010: param error")
                return
            }
            print("MovieReward6010: initAdnetworkSDK")
            strongSelf.amoadInterstitialVideo = AMoAdInterstitialVideo.shared(sid: sid, tag: param.tag)
            strongSelf.amoadInterstitialVideo?.delegate = self
            strongSelf.setCancellable()
            strongSelf.initCompleteAndRetryStartAdIfNeeded()
        }
        
        return true
    }

    override func startAd() -> Bool {
        guard super.startAd() else {
            return false
        }
        
        requireToAsyncRequestAd()

        print("MovieReward6010: startAd")
        if amoadInterstitialVideo?.isLoaded == false {
            amoadInterstitialVideo?.load()
            didLoad = true
        }
        return true
    }

    override func isPrepared() -> Bool {
        guard let amoadInterstitialVideo = amoadInterstitialVideo else {
            return false
        }
        return (didLoad && delegate != nil && amoadInterstitialVideo.isLoaded)
    }

    override func showAd() {
        super.showAd()

        if isPrepared() {
            amoadInterstitialVideo?.show()
            didLoad = false
        } else {
            setPlayFailCallback(PlayFailCallbackReasonIsPreparedFalse, exception: nil)
        }
    }

    override func showAd(withPresenting viewController: UIViewController!) {
        showAd()
    }

    override class func isClassReference() -> Bool {
        return true
    }

    func setCancellable() {
        amoadInterstitialVideo?.isCancellable = false
    }
}

extension MovieReward6010: AMoAdInterstitialVideoDelegate {
    func amoadInterstitialVideoDidLoadAd(amoadInterstitialVideo: AMoAdInterstitialVideo, result: AMoAdResult) {
        print("MovieReward6010: amoadInterstitialVideoDidLoadAd, result : \(result)")
        if result == .success {
            print("MovieReward6010: AMoAdResultSuccess")
            setCallbackStatus(MovieRewardCallbackFetchComplete)
        } else {
            setCallbackStatus(MovieRewardCallbackFetchFail)
        }
    }

    func amoadInterstitialVideoDidStart(amoadInterstitialVideo: AMoAdInterstitialVideo) {
        print("MovieReward6010: amoadInterstitialVideoDidStart")
        setCallbackStatus(MovieRewardCallbackPlayStart)
    }

    func amoadInterstitialVideoDidComplete(amoadInterstitialVideo: AMoAdInterstitialVideo) {
        print("MovieReward6010: amoadInterstitialVideoDidComplete")
        isRewarded = true
        setCallbackStatus(MovieRewardCallbackPlayComplete)
    }

    func amoadInterstitialVideoDidFailToPlay(amoadInterstitialVideo: AMoAdInterstitialVideo) {
        print("MovieReward6010: amoadInterstitialVideoDidFailToPlay")
        setErrorWithMessage("amoadInterstitialVideoDidFailToPlay", code: 0)
        setCallbackStatus(MovieRewardCallbackPlayFail)
    }

    func amoadInterstitialVideoDidShow(amoadInterstitialVideo: AMoAdInterstitialVideo) {
        print("MovieReward6010: amoadInterstitialVideoDidShow")
        setCallbackStatus(MovieRewardCallbackPlayStart)
    }

    func amoadInterstitialVideoWillDismiss(amoadInterstitialVideo: AMoAdInterstitialVideo) {
        print("MovieReward6010: amoadInterstitialVideoWillDismiss")
        setCallbackStatus(MovieRewardCallbackClose)
    }

    func amoadInterstitialVideoDidClickAd(amoadInterstitialVideo: AMoAdInterstitialVideo) {
        print("MovieReward6010: amoadInterstitialVideoDidClickAd")
    }
}

@objc(MovieReward6180)
class MovieReward6180: MovieReward6010 {
}

@objc(MovieReward6181)
class MovieReward6181: MovieReward6010 {
}

@objc(MovieReward6182)
class MovieReward6182: MovieReward6010 {
}

@objc(MovieReward6183)
class MovieReward6183: MovieReward6010 {
}

@objc(MovieReward6184)
class MovieReward6184: MovieReward6010 {
}
